custom/plugins/emspay/Subscriber/PaymentKeeperSubscriber.php

```diff
@@ -1,56 +1,74 @@
-
-        foreach ($args->getSubject()->View()->getAssign()['sPayments'] as $key => $payment) {
-           $name = explode('emspay_',$payment['name'])[1];
-            if ($this->ipAddressValidation($name)) {
-                $allowed_payments[$key] = $payment;
-            }
-
-        }
-        $args->getSubject()->View()->assign('sPayments',$allowed_payments);
-    }
-
-    /**
-     * Check current IP address with Test IP Addresses
-     * @return bool
-     */
-    private function ipAddressValidation($method){
-        $config = Shopware()->Container()->get('shopware.plugin.cached_config_reader')->getByPluginName('emspay');
-        $test_ip = $config[implode('_',['emsonline_test_ip',$method])];
-        if ($test_ip == "") return true;
-        $ip_list = array_map('trim', explode(",", $test_ip));
-
-        return in_array($this->helper->getIpOfTheServer(), $ip_list);
-    }
+<?php
+
+namespace emspay\Subscriber;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Enlight\Event\SubscriberInterface;
+
+class PaymentKeeperSubscriber implements SubscriberInterface
+{
+    protected $ems;
+    protected $helper;
+    protected $config;
+    /**
+     * Subscribe the event
+     * @return array
+     */
+    public static function getSubscribedEvents()
+    {
+
+        return [
+            'Shopware_Controllers_Frontend_Checkout::shippingPaymentAction::after' => 'checkForDisplay',
+        ];
+    }
+
+    /**
+     * Check IP addresses and hide unavailable payment methods
+     * @param \Enlight_Event_EventArgs $args
+     */
+    public function checkForDisplay(\Enlight_Event_EventArgs $args)
+    {
+        $allowed_payments = array();
+        $this->helper = Shopware()->Container()->get('emspay.helper');                                                                          //Create Helper
+        $this->ems = $this->helper->getClient(Shopware()->Container()->get('shopware.plugin.cached_config_reader')->getByPluginName('emspay')); //Create EMS
+        $this->config = Shopware()->Container()->get('shopware.plugin.cached_config_reader')->getByPluginName('emspay');
+        $country = $args->getSubject()->getUserData()['additional']['country']['countryiso'];
+
+        foreach ($args->getSubject()->View()->getAssign()['sPayments'] as $key => $payment) {
+           $name = explode('emspay_',$payment['name'])[1];
+            if ($this->ipAddressValidation($name) && $this->countryValidation($name,$country)) {
+                $allowed_payments[$key] = $payment;
+            }
+
+        }
+        $args->getSubject()->View()->assign('sPayments',$allowed_payments);
+    }
+
+    /**
+     * Check that the user's country matches the country for which Afterpay payment method is available
+     * @param $method
+     * @param $country
+     * @return bool
+     */
+
+    private function countryValidation($method,$country){
+        $countries_string = $this->config[implode('_',['emsonline',$method,'countries'])];
```

Comment:
Could it be that we are accessing an array element by a key that does not exist? For example: $this->config['emsonline_ideal_countries']?

Reply:
there is all okay I have a check later